# nGIA-GP

#### Introduce

This is an alignment-based clustering tool, precise and fast.
It uses GPU and MPI for acceleration and supports both protein and gene sequences.
**It only supports RTX30xx or higher and Linux.**

#### Install

1. Install MPICH.

Install MPICH and copy the library headers to "/usr/local/include", and libraries to "/usr/local/lib". If using the Ubuntu, it can be installed by the package manager.
```bash
sudo apt install mpich
```

2. Install CUDA and GPU Driver.

It can be downloaded from <https://developer.nvidia.com/cuda-downloads>.

3. Compile.

Using make to compile.
```bash
cd makeDB && make && cd ..
cd cluster && make && cd ..
```

#### Usage

1. Make a database.
```bash
cd makeDB && ./makeDB -f ../data/gene.fasta -p ../data/gene.packed -t 0 && cd ..
```
* -f: fasta file (gene or protein sequences)
* -p: packed file (generated by makeDB)
* -t: data type (0-gene 1-protein)

2. Do clustering
```bash
cd cluster && mpirun -n 1 ./cluster -p ../data/gene.packed -r ../data/result.txt -s 0.95 -m 0 && cd ..
```
* -p: packed file (generated by makeDB)
* -r: result file (generated by cluster)
* -s: similarity
* -m: mode (0-fast 1-precise)

#### Result

The generated result file will look like this:
```
1>sequence1
2ACGT
3  >sequence2
4  >sequence3
```
The first and second lines do not have spaces at the beginning, so they are non-redundant sequences. The first line starts with '>', indicating that this is the sequence name, and the second line is the sequence content.
The spaces at the beginning of lines 3 and 4 indicate that these are redundant sequences and are similar to the first line.
Redundant sequences only list sequence names without sequence content.

